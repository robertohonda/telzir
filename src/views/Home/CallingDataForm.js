import React from 'react'
import Grid from '@material-ui/core/Grid'
import TextField from '@material-ui/core/TextField'
import InputAdornment from '@material-ui/core/InputAdornment'
import Button from '@material-ui/core/Button'
import SelectTextField from '../../components/SelectTextField/SelectTextField'

const CallingDataForm = (props) => {
  const {
    values,
    handleChange,
    handleBlur,
    errors,
    origins,
    touched,
    callingRates,
    ratePlans,
    setFieldValue,
    handleSubmit,
    isSubmitting,
    isValid,
    dirty,
  } = props

  const getAvailableDestinations = (origin) =>
    Object.keys(callingRates[origin] == null ? {} : callingRates[origin])

  const onChangeOrigin = (e) => {
    setFieldValue('origin', e.target.value)
    setFieldValue('destination', '')
  }
  const generateOptions = (arrayData) =>
    arrayData.map((value) => ({ label: value, value }))

  return (
    <Grid
      container
      direction="row"
      component="form"
      spacing={2}
      onSubmit={handleSubmit}
    >
      <Grid item xs={12} md={6}>
        <Grid container spacing={2}>
          <Grid item xs={6}>
            <SelectTextField
              label="Origem"
              name="origin"
              fullWidth
              value={values.origin}
              onChange={onChangeOrigin}
              onBlur={handleBlur}
              options={generateOptions(origins)}
              variant="outlined"
            />
          </Grid>
          <Grid item xs={6}>
            <SelectTextField
              label="Destino"
              name="destination"
              fullWidth
              value={values.destination}
              onChange={handleChange}
              onBlur={handleBlur}
              variant="outlined"
              options={generateOptions(getAvailableDestinations(values.origin))}
              disabled={values.origin === ''}
            />
          </Grid>
        </Grid>
      </Grid>
      <Grid item xs={12} md={6}>
        <Grid container alignItems="center" spacing={2}>
          <Grid item xs={6} md={4} lg={5}>
            <TextField
              label="Tempo"
              type="number"
              name="time"
              fullWidth
              value={values.time}
              onChange={handleChange}
              onBlur={handleBlur}
              variant="outlined"
              error={Boolean(touched.time && errors.time)}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">minutos</InputAdornment>
                ),
                inputProps: {
                  min: 0,
                },
              }}
            />
          </Grid>
          <Grid item xs={6} md={4} lg={5}>
            <SelectTextField
              name="ratePlan"
              label="Plano FaleMais"
              fullWidth
              options={ratePlans}
              value={values.ratePlan}
              onChange={handleChange}
              onBlur={handleBlur}
              variant="outlined"
            />
          </Grid>
          <Grid item xs={12} md={4} lg={2}>
            <Button
              type="submit"
              fullWidth
              variant="contained"
              color="primary"
              disabled={!dirty || !isValid || isSubmitting}
            >
              Calcular
            </Button>
          </Grid>
        </Grid>
      </Grid>
    </Grid>
  )
}

export default CallingDataForm
