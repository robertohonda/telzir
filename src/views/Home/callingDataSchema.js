import * as yup from 'yup'

const callingDataSchema = yup.object().shape({
  origin: yup.string().required(),
  destination: yup.string().required(),
  time: yup
    .number()
    .positive()
    .integer()
    .required(),
  ratePlan: yup.string().required(),
})

export default callingDataSchema
