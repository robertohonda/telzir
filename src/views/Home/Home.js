import { makeStyles } from '@material-ui/core/styles'
import React from 'react'
import CostCard from '../../components/CostCard/CostCard'
import Grid from '@material-ui/core/Grid'
import { callingRates, ratePlans, ratePlansFreeMinutes } from '../../data'
import { connect } from 'react-redux'
import {
  setCallingData,
  setCallingDataLoading,
} from '../../redux/actions/callingData'

import { Formik } from 'formik'
import CallingDataForm from './CallingDataForm'
import callingDataSchema from './callingDataSchema'

const useStyles = makeStyles((theme) => ({
  priceSection: {
    marginTop: theme.spacing(3),
  },
}))

export const calculateCostWithRatePlan = (
  origin,
  destination,
  time,
  ratePlan,
) => {
  const tax = callingRates[origin][destination].cost
  const freeMinutes = ratePlansFreeMinutes[ratePlan]

  return Math.max(time - freeMinutes, 0) * 1.1 * tax
}

export const calculateCostWithoutRatePlan = (origin, destination, time) => {
  const tax = callingRates[origin][destination].cost
  return time * tax
}

const Home = (props) => {
  const classes = useStyles()
  const { callingData, setCallingData, setCallingDataLoading } = props
  const origins = Object.keys(callingRates)

  const onSubmitCalculate = (callingData, { setSubmitting }) => {
    const { origin, destination, time, ratePlan } = callingData
    const costWithRatePlan = calculateCostWithRatePlan(
      origin,
      destination,
      time,
      ratePlan,
    )
    const costWithoutRatePlan = calculateCostWithoutRatePlan(
      origin,
      destination,
      time,
    )
    setCallingDataLoading(true)
    setTimeout(() => {
      setCallingData({
        ...callingData,
        costWithoutRatePlan,
        costWithRatePlan,
      })
      setSubmitting(false)
    }, 1000)
  }

  const inputsSection = (
    <Formik
      onSubmit={onSubmitCalculate}
      validationSchema={callingDataSchema}
      initialValues={callingData}
      render={(props) => (
        <CallingDataForm
          origins={origins}
          callingRates={callingRates}
          ratePlans={ratePlans}
          {...props}
        />
      )}
    />
  )

  const priceSection = (
    <Grid container justify="center" spacing={3}>
      <Grid item xs={12} sm={6} md={4} lg={3}>
        <CostCard
          loading={callingData.loading}
          title="Com FaleMais"
          cost={
            callingData.costWithRatePlan == null
              ? '- -'
              : `$ ${callingData.costWithRatePlan.toFixed(2)}`
          }
        />
      </Grid>
      <Grid item xs={12} sm={6} md={4} lg={3}>
        <CostCard
          loading={callingData.loading}
          title="Sem FaleMais"
          cost={
            callingData.costWithoutRatePlan == null
              ? '- -'
              : `$ ${callingData.costWithoutRatePlan.toFixed(2)}`
          }
        />
      </Grid>
    </Grid>
  )

  return (
    <div>
      <div>{inputsSection}</div>
      <div className={classes.priceSection}>{priceSection}</div>
    </div>
  )
}

const mapStateToProps = ({ callingData }) => ({ callingData })

export default connect(
  mapStateToProps,
  {
    setCallingData,
    setCallingDataLoading,
  },
)(Home)
