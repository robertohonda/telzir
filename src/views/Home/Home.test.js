import React from 'react'
import { shallow } from 'enzyme'
import Home, {
  calculateCostWithRatePlan,
  calculateCostWithoutRatePlan,
} from './Home'
import configureStore from 'redux-mock-store'

const store = configureStore([])()

describe('Home', () => {
  it('should render correctly', () => {
    const component = shallow(<Home store={store} />)

    expect(component).toMatchSnapshot()
  })

  it('should calculate the cost correctly', () => {
    expect(calculateCostWithRatePlan('011', '016', 20, 'FALEMAIS30')).toEqual(0)
    expect(calculateCostWithoutRatePlan('011', '016', 20)).toEqual(38)
    expect(calculateCostWithRatePlan('011', '017', 80, 'FALEMAIS60')).toEqual(
      37.4,
    )
    expect(calculateCostWithoutRatePlan('011', '017', 80)).toEqual(136.0)
    expect(calculateCostWithRatePlan('018', '011', 200, 'FALEMAIS120')).toEqual(
      167.2,
    )
    expect(calculateCostWithoutRatePlan('018', '011', 200)).toEqual(380.0)
  })
})
