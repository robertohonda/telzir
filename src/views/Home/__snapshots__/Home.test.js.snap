// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<ContextProvider
  value={null}
>
  <Home
    setCallingData={[Function]}
    setCallingDataLoading={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
